internal/data/generated_metrics.go

```diff
@@ -120,9 +121,9 @@ func (ms ResourceMetrics) Resource() Resource {
 	return newResource(ms.orig.Resource)
 }
 
-// SetResource replaces the resource associated with this ResourceMetrics.
-func (ms ResourceMetrics) SetResource(v Resource) {
-	ms.orig.Resource = v.orig
+// HasResource returns true if there is a resource associated with this ResourceMetrics.
+func (ms ResourceMetrics) HasResource() bool {
```

Comment:
The semantics are a bit unclear. For a newly consturcted `ResourceMetrics`  `HasResource()` will return false. But if I call `Resource()` to attempt to get data and after that call `HasResource()` it will return true. It is a bit surprising behavior.

Perhaps we should make `HasResource()` stronger? e.g.:

`return ms.orig.Resource != nil && *ms.orig.Resource!=otlpresource.Resource{}`

This will ensure `HasResource()` returns for false for default initialized `Resource`, which solves the above problem.

Reply:
What about this:
```
func (ms ResourceMetrics) HasResource() bool {
return ms.orig.Resource != nil && !newResoruce(ms.orig.Resource).IsDefault()
}

func (ms Resource) IsDefault() bool {
*ms.orig.Resource==otlpresource.Resource{}
}
```

This way you can check even elements in a slice if they are default.

Reply:
> ```
> func (ms ResourceMetrics) HasResource() bool {
> return ms.orig.Resource != nil && !newResoruce(ms.orig.Resource).IsDefault()
> }
>```

This will be equivalent to `ms.orig.Resource != nil && *ms.orig.Resource!=otlpresource.Resource{}`, right?


> ```
> func (ms Resource) IsDefault() bool {
> *ms.orig == otlpresource.Resource{}
> }
> ```
> 
> This way you can check even elements in a slice if they are default.

Sounds good.

Reply:
> This will be equivalent to ms.orig.Resource != nil && *ms.orig.Resource!=otlpresource.Resource{}, right?

Correct

Reply:
SGTM.

Reply:
As discussed offline we went with a bit of a different approach mostly because "==" operation is not defined for slices and some proto generated structs have slices embedded.

```
// Getter no init.
func (ms ResourceMetrics) Resource() {
  newResoruce(ms.orig.Resource)
}
// Initialize with empty if nil
func (ms ResourceMetrics) InitResourceIfNil() bool {
  if ms.orig.Resource == nil {
    ms.orig.Resource = &ms.orig.Resource{}
  }
}

func (ms Resource) IsNil() bool {
  ms.orig == otlpresource.Resource{}
}

// All methods except IsNil will crash if IsNil returns true.
```

Reply:
@tigrannajaryan I think it is possible if we use double pointers in the Resource to move InitIfNil on the resource if that is nicer. That will improve `func ocStatusToInternal(ocStatus *octrace.Status, out data.Span)` to accept the `out data.SpanStatus` and not have to pass the parent.

Reply:
With double pointers I am not sure how we will avoid extra allocations. Maybe there is a way, let's see it in the code.

Reply:
Nobody needs to call the NewEmptyFoo constructor except in tests